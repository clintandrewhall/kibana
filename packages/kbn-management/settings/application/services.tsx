/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { FC } from 'react';

import {
  FieldRowProvider,
  FieldRowKibanaProvider,
  type FieldRowKibanaDependencies,
  type FieldRowServices,
} from '@kbn/management-settings-components-field-row';

export type SettingsApplicationServices = FieldRowServices;
export type SettingsApplicationKibanaDependencies = FieldRowKibanaDependencies;

/**
 * A Context Provider that provides services to the component and its dependencies.
 */
export const SettingsApplicationProvider: FC<SettingsApplicationServices> = ({
  children,
  ...services
}) => {
  return <FieldRowProvider {...services}>{children}</FieldRowProvider>;
};

/**
 * Kibana-specific Provider that maps dependencies to services.
 */
export const SettingsApplicationKibanaProvider: FC<SettingsApplicationKibanaDependencies> = ({
  children,
  ...dependencies
}) => {
  return <FieldRowKibanaProvider {...dependencies}>{children}</FieldRowKibanaProvider>;
};
