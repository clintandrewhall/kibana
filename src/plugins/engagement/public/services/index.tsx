/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { FC, createContext, useContext } from 'react';

interface WithoutChat {
  enabled: false;
}

interface WithChat {
  enabled: true;
  chatURL: string;
  identityJWT: string;
  userID: string;
  userEmail: string;
}

type ChatService = WithChat | WithoutChat;

export interface EngagementServices {
  chat: ChatService;
}

const ServicesContext = createContext<EngagementServices>({ chat: { enabled: false } });

export const ServicesProvider: FC<EngagementServices> = ({ children, ...services }) => (
  <ServicesContext.Provider value={services}>{children}</ServicesContext.Provider>
);

/**
 * React hook for accessing the pre-wired `EngagementServices`.
 */
export function useServices() {
  return useContext(ServicesContext);
}

export function useChat(): ChatService {
  const { chat } = useServices();
  return chat;
}
