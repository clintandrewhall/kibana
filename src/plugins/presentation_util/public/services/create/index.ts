/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

import { mapValues } from 'lodash';

import { PluginServiceRegistry } from './registry';

export { PluginServiceRegistry } from './registry';
export { PluginServiceProvider, PluginServiceProviders } from './provider';
export {
  PluginServiceFactory,
  KibanaPluginServiceFactory,
  KibanaPluginServiceParams,
} from './factory';

/**
 * `PluginServices` is a top-level class for specifying and accessing services within a plugin.
 *
 * A `PluginServices` object can be provided with a `PluginServiceRegistry` at any time, which will
 * then be used to provide services to any component that accesses it.
 *
 * The `Services` generic determines the shape of all service APIs being produced.
 */
export class PluginServices<Services> {
  private registry: PluginServiceRegistry<Services, any> | null = null;

  setRegistry(registry: PluginServiceRegistry<Services, any> | null) {
    if (registry && !registry.isStarted()) {
      throw new Error('Registry has not been started.');
    }

    this.registry = registry;
  }

  hasRegistry() {
    return !!this.registry;
  }

  private getRegistry() {
    if (!this.registry) {
      throw new Error('No registry has been provided.');
    }

    return this.registry;
  }

  getContextProvider() {
    return this.getRegistry().getContextProvider();
  }

  getContextHooks(): { [K in keyof Services]: { useContext: () => Services[K] } } {
    const registry = this.getRegistry();
    const providers = registry.getServiceProviders();

    // @ts-expect-error Need to fix this; the type isn't fully understood inferred.
    return mapValues(providers, (provider) => ({
      useContext: provider.useContext(),
    }));
  }
}
