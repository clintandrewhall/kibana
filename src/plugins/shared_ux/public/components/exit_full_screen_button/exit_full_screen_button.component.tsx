/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { MouseEventHandler, HTMLAttributes } from 'react';
import {
  EuiScreenReaderOnly,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiText,
  useEuiTheme,
  makeHighContrastColor,
} from '@elastic/eui';
import { i18n } from '@kbn/i18n';
import { css } from '@emotion/react';
import cx from 'classnames';

import './exit_full_screen_button.scss';

const label = i18n.translate('sharedUX.exitFullScreenButton.exitFullScreenModeButtonAriaLabel', {
  defaultMessage: 'Exit full screen mode',
});

const text = i18n.translate('sharedUX.exitFullScreenButton.exitFullScreenModeButtonText', {
  defaultMessage: 'Exit full screen',
});

const description = i18n.translate('sharedUX.exitFullScreenButton.fullScreenModeDescription', {
  defaultMessage: 'In full screen mode, press ESC to exit.',
});

/**
 * Props for the Exit Full Screen button component.
 */
export interface Props extends Pick<HTMLAttributes<HTMLDivElement>, 'className'> {
  onClick: MouseEventHandler<HTMLButtonElement>;
}

/**
 * A presentational component that renders a button designed to exit "full screen" mode.
 */
export const ExitFullScreenButton = ({ onClick, className }: Props) => {
  const { euiTheme } = useEuiTheme();
  const { colors, size, border } = euiTheme;

  const textCSS = css`
    line-height: 1.2;
    ${makeHighContrastColor(colors.mediumShade)(colors.fullShade)};
  `;

  const iconCSS = css`
    ${makeHighContrastColor(colors.mediumShade)(colors.fullShade)};
  `;

  const buttonCSS = css`
    display: inline-block;
    padding: ${size.xs} ${size.s};
    border: none;
    background: none;
    background: ${colors.fullShade};
    border-radius: ${border.radius.small};
    text-align: left;
    height: 32px;
    color: ${colors.lightShade};

    &:hover {
      background: ${colors.fullShade};
      color: ${colors.emptyShade};
    }
  `;

  return (
    <div css={buttonCSS} className={cx('exitFullScreenButton', className)}>
      <EuiScreenReaderOnly>
        <p aria-live="polite">{description}</p>
      </EuiScreenReaderOnly>
      <button aria-label={label} onClick={onClick}>
        <EuiFlexGroup component="span" responsive={false} alignItems="center" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiIcon type="logoElastic" size="l" />
          </EuiFlexItem>
          <EuiFlexItem grow={false} data-test-subj="exitFullScreenModeText">
            <EuiText size="s" css={textCSS}>
              {text}
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiIcon type="fullScreen" css={iconCSS} />
          </EuiFlexItem>
        </EuiFlexGroup>
      </button>
    </div>
  );
};
