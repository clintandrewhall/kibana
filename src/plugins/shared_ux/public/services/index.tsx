/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { FC, createContext, useContext } from 'react';
import { SharedUXPlatformService } from './platform';
import { servicesFactory } from './stub';

export interface SharedUXServices {
  platform: SharedUXPlatformService;
}

const ServicesContext = createContext<SharedUXServices>(servicesFactory());

export const ServicesProvider: FC<SharedUXServices> = ({ children, ...services }) => (
  <ServicesContext.Provider value={services}>{children}</ServicesContext.Provider>
);

export function useServices() {
  return useContext(ServicesContext);
}

export const usePlatformService = () => useServices().platform;
