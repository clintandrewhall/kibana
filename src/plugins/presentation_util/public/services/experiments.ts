/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { IUiSettingsClient } from 'kibana/public';
import {
  ExperimentEnvironment,
  experimentIDs,
  Experiment,
  ExperimentConfig,
  ExperimentID,
  EnvironmentStatus,
  environmentNames,
  isExperimentEnabledByStatus,
} from '../../common';

export interface PresentationExperimentsService {
  getExperimentIDs: () => typeof experimentIDs;
  getExperiment: (id: ExperimentID) => Experiment;
  getExperiments: () => Record<ExperimentID, Experiment>;
  setExperimentStatus: (id: ExperimentID, env: ExperimentEnvironment, status: boolean) => void;
  reset: () => void;
}

export const isEnabledByStorageValue = (
  experiment: ExperimentConfig,
  environment: ExperimentEnvironment,
  value: string | boolean | null
): boolean => {
  const defaultValue = experiment.isActive;

  if (!experiment.environments.includes(environment)) {
    return defaultValue;
  }

  if (value === true || value === false) {
    return value;
  }

  if (value === 'enabled') {
    return true;
  }

  if (value === 'disabled') {
    return false;
  }

  return defaultValue;
};

export const setStorageStatus = (storage: Storage, id: ExperimentID, enabled: boolean) =>
  storage.setItem(id, enabled ? 'enabled' : 'disabled');

export const applyExperimentStatus = (
  experiment: ExperimentConfig,
  status: EnvironmentStatus
): Experiment => {
  const { isActive, environments } = experiment;

  environmentNames.forEach((name) => {
    if (!environments.includes(name)) {
      delete status[name];
    }
  });

  const isEnabled = isExperimentEnabledByStatus(isActive, status);
  const isOverride = isEnabled !== isActive;

  return {
    ...experiment,
    status: {
      ...status,
      defaultValue: isActive,
      isEnabled,
      isOverride,
    },
  };
};

export const setUISettingsStatus = (
  client: IUiSettingsClient,
  id: ExperimentID,
  enabled: boolean
) => client.set(id, enabled);
