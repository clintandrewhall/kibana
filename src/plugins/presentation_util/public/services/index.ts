/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { PluginServices } from './create';
import { PresentationCapabilitiesService } from './capabilities';
import { PresentationDashboardsService } from './dashboards';
import { PresentationExperimentsService } from './experiments';
export interface PresentationUtilServices {
  dashboards: PresentationDashboardsService;
  capabilities: PresentationCapabilitiesService;
  experiments: PresentationExperimentsService;
}

export const pluginServices = new PluginServices<PresentationUtilServices>();
