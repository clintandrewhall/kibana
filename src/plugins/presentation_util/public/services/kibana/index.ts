/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { capabilitiesServiceFactory } from './capabilities';
import { dashboardsServiceFactory } from './dashboards';
import { experimentsServiceFactory } from './experiments';
import {
  PluginServiceProviders,
  KibanaPluginServiceParams,
  PluginServiceProvider,
  PluginServiceRegistry,
} from '../create';
import { PresentationUtilPluginStartDeps } from '../../types';
import { PresentationUtilServices } from '..';

export { capabilitiesServiceFactory } from './capabilities';
export { dashboardsServiceFactory } from './dashboards';
export { experimentsServiceFactory } from './experiments';

export const providers: PluginServiceProviders<
  PresentationUtilServices,
  KibanaPluginServiceParams<PresentationUtilPluginStartDeps>
> = {
  capabilities: new PluginServiceProvider(capabilitiesServiceFactory),
  experiments: new PluginServiceProvider(experimentsServiceFactory),
  dashboards: new PluginServiceProvider(dashboardsServiceFactory),
};

export const registry = new PluginServiceRegistry<
  PresentationUtilServices,
  KibanaPluginServiceParams<PresentationUtilPluginStartDeps>
>(providers);
