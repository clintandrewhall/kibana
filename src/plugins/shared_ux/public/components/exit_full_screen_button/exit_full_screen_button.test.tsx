/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { FC } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { keys } from '@elastic/eui';

import { ServicesProvider } from '../../services';
import { servicesFactory } from '../../services/jest';
import { ExitFullScreenButton } from './exit_full_screen_button';

const services = servicesFactory();
const Context: FC = ({ children }) => <ServicesProvider {...services}>{children}</ServicesProvider>;

describe('<ExitFullScreenButton />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('is rendered', () => {
    const component = mount(
      <Context>
        <ExitFullScreenButton onExit={jest.fn()} />
      </Context>
    );

    expect(component).toMatchSnapshot();
  });

  test('passing `false` to toggleChrome does not toggle chrome', () => {
    const component = mount(
      <Context>
        <ExitFullScreenButton onExit={jest.fn()} toggleChrome={false} />
      </Context>
    );
    expect(services.chrome.setIsChromeVisible).toHaveBeenCalledTimes(0);
    component.unmount();
    expect(services.chrome.setIsChromeVisible).toHaveBeenCalledTimes(0);
  });

  describe('onExit', () => {
    const onExitHandler = jest.fn();
    let component: ReactWrapper;

    beforeEach(() => {
      component = mount(
        <Context>
          <ExitFullScreenButton onExit={onExitHandler} toggleChrome={true} />
        </Context>
      );
    });

    test('is called when the button is pressed', () => {
      expect(services.chrome.setIsChromeVisible).toHaveBeenLastCalledWith(false);

      component.find('button').simulate('click');

      expect(onExitHandler).toHaveBeenCalledTimes(1);
      component.unmount();
      expect(services.chrome.setIsChromeVisible).toHaveBeenLastCalledWith(true);
    });

    test('is called when the ESC key is pressed', () => {
      expect(services.chrome.setIsChromeVisible).toHaveBeenLastCalledWith(false);

      const escapeKeyEvent = new KeyboardEvent('keydown', { key: keys.ESCAPE } as any);
      document.dispatchEvent(escapeKeyEvent);

      expect(onExitHandler).toHaveBeenCalledTimes(1);
      component.unmount();
      expect(services.chrome.setIsChromeVisible).toHaveBeenLastCalledWith(true);
    });
  });
});
