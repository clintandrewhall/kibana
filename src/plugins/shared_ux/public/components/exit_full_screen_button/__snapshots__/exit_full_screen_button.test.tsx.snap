// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExitFullScreenButton /> is rendered 1`] = `
.emotion-0 {
  line-height: 1.2;
}

<ServicesProvider
  platform={
    Object {
      "setIsFullscreen": [MockFunction],
    }
  }
>
  <ExitFullScreenButton
    onExit={[MockFunction]}
  >
    <ExitFullScreenButton
      css={
        Object {
          "map": undefined,
          "name": "knjqgf",
          "next": undefined,
          "styles": "
    bottom: 8px;
    left: 8px;
    position: fixed;
    z-index: 5;
  ",
          "toString": [Function],
        }
      }
      onClick={[Function]}
    >
      <div
        className="exitFullScreenButton"
      >
        <EuiScreenReaderOnly>
          <p
            aria-live="polite"
            className="euiScreenReaderOnly"
          >
            In full screen mode, press ESC to exit.
          </p>
        </EuiScreenReaderOnly>
        <button
          aria-label="Exit full screen mode"
          onClick={[Function]}
        >
          <EuiFlexGroup
            alignItems="center"
            component="span"
            gutterSize="s"
            responsive={false}
          >
            <span
              className="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
            >
              <EuiFlexItem
                grow={false}
              >
                <div
                  className="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <EuiIcon
                    size="l"
                    type="logoElastic"
                  >
                    <span
                      data-euiicon-type="logoElastic"
                      size="l"
                    />
                  </EuiIcon>
                </div>
              </EuiFlexItem>
              <EuiFlexItem
                data-test-subj="exitFullScreenModeText"
                grow={false}
              >
                <div
                  className="euiFlexItem euiFlexItem--flexGrowZero"
                  data-test-subj="exitFullScreenModeText"
                >
                  <EuiText
                    css={
                      Object {
                        "map": undefined,
                        "name": "bouowd",
                        "next": undefined,
                        "styles": "
    line-height: 1.2;
    #98a2b3;
  ",
                        "toString": [Function],
                      }
                    }
                    size="s"
                  >
                    <div
                      css="unknown styles"
                    >
                      <div
                        className="euiText euiText--small emotion-0"
                      >
                        Exit full screen
                      </div>
                    </div>
                  </EuiText>
                </div>
              </EuiFlexItem>
              <EuiFlexItem
                grow={false}
              >
                <div
                  className="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <EuiIcon
                    css={
                      Object {
                        "map": undefined,
                        "name": "1xgac5z",
                        "next": undefined,
                        "styles": "
    #98a2b3;
  ",
                        "toString": [Function],
                      }
                    }
                    type="fullScreen"
                  >
                    <span
                      css="unknown styles"
                      data-euiicon-type="fullScreen"
                    >
                      <span
                        className="emotion-1"
                        data-euiicon-type="fullScreen"
                      />
                    </span>
                  </EuiIcon>
                </div>
              </EuiFlexItem>
            </span>
          </EuiFlexGroup>
        </button>
      </div>
    </ExitFullScreenButton>
  </ExitFullScreenButton>
</ServicesProvider>
`;
